third_party/nvfuser/csrc/expr_simplifier.cpp

```diff
@@ -1105,7 +1105,16 @@ bool isPositive(Val* value, const VarInfoMap& var_info) {
     return true;
   }
   if (auto ns = dynamic_cast<NamedScalar*>(value)) {
-    if (ns->getParallelDim().has_value()) {
+    // TODO: currently we are implicitly assuming tensor sizes to be non-zero.
+    // For example, if I have T0[2, 0] and I do T0->merge(0), I get T0[2*0]. And
```

Comment:
Does this happen when that tensor is contiguous? Can you create a repro?

Reply:
Yes, this do happen, but not on the most trivial case. For example, in `FusionCodeGen_CUDA`, the fusion is just the most simple pointwise kernel (and there is no broadcasting). The schedule is:
```C++
  //[I0, I1, I2]
  tv2 = tv2->split(0, 4);
  //[I0o, I0i{4}, I1, I2]
  tv2 = tv2->merge(1);
  //[I0o, I0i{4}*I1, I2]
  tv2 = tv2->split(-1, 2);
  //[I0o, I0i{4}*I1, I2o, I2i{2}]
  tv2 = tv2->reorder({{0, 1}, {1, 0}, {3, 2}});
  //[I0i{4}*I1, I0o, I2i{2}, I2o]
```
If we look at the tensor `tv2` (which is contiguous), the first dimension is `I0i{4}*I1`, it is a merge of the inner part of `I0` with `I1`, so it is a contiguous merge. So we would expect that the index wrt the first dimension (in this case `i31`) to be linear and have stride `I2`, so the `tv2` should be accessed with something like `T2[i31 * T2.size[2] + ...]`. However, without this pass, the index is not like that. See the following print:

**Disable this pass:**
```
PYTORCH_NVFUSER_DISABLE="index_hoist,expr_simplify(fundamentalDivisionWithRemainderProperty)" PYTORCH_NVFUSER_DUMP="expr_simplify(fundamentalDivisionWithRemainderProperty),cuda_kernel" ./build/bin/nvfuser_tests --gtest_filter=*FusionCodeGen_CUDA*
```

```C++
======= Codegen output for kernel: kernel1 =======

__global__ void kernel1(Tensor<float, 3> T2) {
  NVFUSER_DEFINE_MAGIC_ZERO
  #pragma unroll 1
  for(nvfuser_index_t i31 = 0; i31 < (4 * T2.size[1]); ++i31) {
    #pragma unroll 1
    for(nvfuser_index_t i32 = 0; i32 < (ceilDiv(T2.size[0], 4)); ++i32) {
      #pragma unroll
      for(nvfuser_index_t i33 = 0; i33 < 2; ++i33) {
        #pragma unroll 1
        for(nvfuser_index_t i34 = 0; i34 < (ceilDiv(T2.size[2], 2)); ++i34) {
          float T0[1];
          T0[0]
            = 0.00000000000000000e+00
            + 1.00000000000000000e+00;
          float T1[1];
          T1[0]
            = T0[0]
            + (float) 2.00000000000000000e+00;
          if ((((0 < (T2.size[1] + (-(i31 % T2.size[1])))) && (0 < ((T2.size[0] + (-(i31 / T2.size[1]))) + (-(4 * i32))))) && (0 < ((T2.size[2] + (-(i33 + nvfuser_zero))) + (-(2 * i34)))))) {
            T2[((((((T2.size[2] * T2.size[1]) * (i31 / T2.size[1])) + (T2.size[2] * (i31 % T2.size[1]))) + (((T2.size[2] * T2.size[1]) * 4) * i32)) + (i33 + nvfuser_zero)) + (2 * i34))]
              = T1[0]
              + (float) 3.00000000000000000e+00;
          }
        }
      }
      NVFUSER_UPDATE_MAGIC_ZERO
    }
  }
}

======================================
```

**Enable this pass:**
```
PYTORCH_NVFUSER_DISABLE="index_hoist" PYTORCH_NVFUSER_DUMP="expr_simplify(fundamentalDivisionWithRemainderProperty),cuda_kernel" ./build/bin/nvfuser_tests --gtest_filter=*FusionCodeGen_CUDA*
```
```C++
Simplifying expression:
i82 = ( ( ( ( ( i32 * 4 ) + ( i31 / T2.size[1] ) ) * ( T2.size[2] * T2.size[1] ) ) + ( ( i31 % T2.size[1] ) * T2.size[2] ) ) + ( ( i34 * 2 ) + ( i33 + nvfuser_zero ) ) )
assoc_comm::flatten:
i93 = ( FlattenedAdd(( FlattenedMul(( FlattenedAdd(( FlattenedMul(i32, 4) ), ( i31 / T2.size[1] )) ), T2.size[2], T2.size[1]) ), ( FlattenedMul(( i31 % T2.size[1] ), T2.size[2]) ), ( FlattenedMul(i34, 2) ), ( i33 + nvfuser_zero )) )
distributeMul:
i101 = ( FlattenedAdd(( FlattenedAdd(( FlattenedMul(T2.size[2], T2.size[1], ( FlattenedMul(i32, 4) )) ), ( FlattenedMul(T2.size[2], T2.size[1], ( i31 / T2.size[1] )) )) ), ( FlattenedMul(( i31 % T2.size[1] ), T2.size[2]) ), ( FlattenedMul(i34, 2) ), ( i33 + nvfuser_zero )) )
assoc_comm::flatten:
i104 = ( FlattenedAdd(( FlattenedMul(T2.size[2], T2.size[1], i32, 4) ), ( FlattenedMul(T2.size[2], T2.size[1], ( i31 / T2.size[1] )) ), ( FlattenedMul(( i31 % T2.size[1] ), T2.size[2]) ), ( FlattenedMul(i34, 2) ), ( i33 + nvfuser_zero )) )
fundamentalDivisionWithRemainderProperty:
i109 = ( FlattenedAdd(( FlattenedMul(i31, T2.size[2]) ), ( FlattenedMul(T2.size[2], T2.size[1], i32, 4) ), ( FlattenedMul(i34, 2) ), ( i33 + nvfuser_zero )) )
assoc_comm::unflatten:
i117 = ( ( ( ( T2.size[2] * i31 ) + ( ( ( T2.size[2] * T2.size[1] ) * 4 ) * i32 ) ) + ( i33 + nvfuser_zero ) ) + ( 2 * i34 ) )
================================================================================

======= Codegen output for kernel: kernel1 =======

__global__ void kernel1(Tensor<float, 3> T2) {
  NVFUSER_DEFINE_MAGIC_ZERO
  #pragma unroll 1
  for(nvfuser_index_t i31 = 0; i31 < (4 * T2.size[1]); ++i31) {
    #pragma unroll 1
    for(nvfuser_index_t i32 = 0; i32 < (ceilDiv(T2.size[0], 4)); ++i32) {
      #pragma unroll
      for(nvfuser_index_t i33 = 0; i33 < 2; ++i33) {
        #pragma unroll 1
        for(nvfuser_index_t i34 = 0; i34 < (ceilDiv(T2.size[2], 2)); ++i34) {
          float T0[1];
          T0[0]
            = 0.00000000000000000e+00
            + 1.00000000000000000e+00;
          float T1[1];
          T1[0]
            = T0[0]
            + (float) 2.00000000000000000e+00;
          if ((((0 < (T2.size[1] + (-(i31 % T2.size[1])))) && (0 < ((T2.size[0] + (-(i31 / T2.size[1]))) + (-(4 * i32))))) && (0 < ((T2.size[2] + (-(i33 + nvfuser_zero))) + (-(2 * i34)))))) {
            T2[((((T2.size[2] * i31) + (((T2.size[2] * T2.size[1]) * 4) * i32)) + (i33 + nvfuser_zero)) + (2 * i34))]
              = T1[0]
              + (float) 3.00000000000000000e+00;
          }
        }
      }
      NVFUSER_UPDATE_MAGIC_ZERO
    }
  }
}

======================================
```